File: sahaj2.java
import java.util.*;

public class sahaj2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int num = scanner.nextInt();
        scanner.nextLine();
        String[] arr_words = scanner.nextLine().split(" ");
        String req = scanner.nextLine().trim();
        Set<String> word_set = new HashSet<>(Arrays.asList(arr_words));

        if (solve(req, word_set)) {
            System.out.println("Harry");
        } else {
            System.out.println("Voldemort");
        }
    }

    public static boolean solve(String word, Set<String> word_set) {
        if (word.length() == 0) {
            return true;
        }
        for (int i = 1; i <= word.length(); i++) {
            String pre = word.substring(0, i);
            if (word_set.contains(pre) && solve(word.substring(i), word_set)) {
                return true;
            }
        }
        return false;
    }
}

File: sahaj2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 100

int solve(char *word, char **word_set, int word_set_size);

int main() {
    int num;
    scanf("%d", &num);

    char **arr_words = (char **)malloc(num * sizeof(char *));
    for (int i = 0; i < num; i++) {
        arr_words[i] = (char *)malloc(MAX_LENGTH * sizeof(char));
        scanf("%s", arr_words[i]);
    }

    char req[MAX_LENGTH];
    scanf("%s", req);

    char **word_set = (char **)malloc(num * sizeof(char *));
    int word_set_size = 0;
    for (int i = 0; i < num; i++) {
        if (strcmp(arr_words[i], req) == 0) {
            continue;
        }
        word_set[word_set_size] = (char *)malloc(MAX_LENGTH * sizeof(char));
        strcpy(word_set[word_set_size], arr_words[i]);
        word_set_size++;
    }

    if (solve(req, word_set, word_set_size)) {
        printf("Harry\n");
    } else {
        printf("Voldemort\n");
    }

    for (int i = 0; i < num; i++) {
        free(arr_words[i]);
    }
    free(arr_words);

    for (int i = 0; i < word_set_size; i++) {
        free(word_set[i]);
    }
    free(word_set);

    return 0;
}

int solve(char *word, char **word_set, int word_set_size) {
    if (strlen(word) == 0) {
        return 1;
    }
    for (int i = 1; i <= strlen(word); i++) {
        char pre[MAX_LENGTH];
        strncpy(pre, word, i);
        pre[i] = '\0';
        for (int j = 0; j < word_set_size; j++) {
            if (strcmp(pre, word_set[j]) == 0 && solve(word + i, word_set, word_set_size)) {
                return 1;
            }
        }
    }
    return 0;
}
